Can't run tests on stable or without --all-features

This makes contributing very difficult. I expected `cargo test` to work. After digging into `.travis.yml` I figured that `cargo test --all-features` is what I need, but that doesn't work on stable. It also looks like tests aren't run against stable in CI?

I agree that there isn't going to be a way for us to run all tests on stable but I sympathize that this can make contributing tricky. Let's use [`rustc_version`](https://crates.io/crates/rustc_version) in a build script to detect when we are testing on nightly, otherwise replace the precedence and round-trip tests with a single #[test] that unconditionally panics with a message that says not all tests ran with an explanation. That way that rest of the tests actually run and you get something more helpful than:

```
error[E0554]: #![feature] may not be used on the stable release channel
  --> tests/test_round_trip.rs:10:1
   |
10 | #![feature(rustc_private)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -39,6 +39,7 @@ quote = { version = "0.6", optional = true, default-features = false }
 unicode-xid = "0.1"
 
 [dev-dependencies]
+colored = "1.7"
 rayon = "1.0"
 regex = "1.0"
 walkdir = "2.1"

--- a/build.rs
+++ b/build.rs
@@ -6,22 +6,31 @@ use std::str::{self, FromStr};
 // opening a GitHub issue if your build environment requires some way to enable
 // these cfgs other than by executing our build script.
 fn main() {
-    let minor = match rustc_minor_version() {
-        Some(minor) => minor,
+    let compiler = match rustc_version() {
+        Some(compiler) => compiler,
         None => return,
     };
 
-    if minor >= 19 {
+    if compiler.minor >= 19 {
         println!("cargo:rustc-cfg=syn_can_use_thread_id");
     }
 
     // Macro modularization allows re-exporting the `quote!` macro in 1.30+.
-    if minor >= 30 {
+    if compiler.minor >= 30 {
         println!("cargo:rustc-cfg=syn_can_call_macro_by_path");
+    }
+
+    if !compiler.nightly {
+        println!("cargo:rustc-cfg=syn_disable_nightly_tests");
     }
 }
 
-fn rustc_minor_version() -> Option<u32> {
+struct Compiler {
+    minor: u32,
+    nightly: bool,
+}
+
+fn rustc_version() -> Option<Compiler> {
     let rustc = match env::var_os("RUSTC") {
         Some(rustc) => rustc,
         None => return None,
@@ -47,5 +56,13 @@ fn rustc_minor_version() -> Option<u32> {
         None => return None,
     };
 
-    u32::from_str(next).ok()
+    let minor = match u32::from_str(next) {
+        Ok(minor) => minor,
+        Err(_) => return None,
+    };
+
+    Some(Compiler {
+        minor: minor,
+        nightly: version.contains("nightly"),
+    })
 }

--- a/tests/test_generics.rs
+++ b/tests/test_generics.rs
@@ -1,6 +1,4 @@
-#![cfg(not(syn_disable_nightly_tests))]
 #![recursion_limit = "1024"]
-#![feature(rustc_private)]
 
 extern crate syn;
 use syn::*;
@@ -14,7 +12,6 @@ use proc_macro2::{Ident, Span, TokenStream};
 #[macro_use]
 mod macros;
 
-mod common;
 mod features;
 
 fn ident(s: &str) -> Ident {

--- a/tests/test_generics.rs
+++ b/tests/test_generics.rs
@@ -1,3 +1,4 @@
+#![cfg(not(syn_disable_nightly_tests))]
 #![recursion_limit = "1024"]
 #![feature(rustc_private)]
 

--- a/tests/test_grouping.rs
+++ b/tests/test_grouping.rs
@@ -1,6 +1,4 @@
-#![cfg(not(syn_disable_nightly_tests))]
 #![recursion_limit = "1024"]
-#![feature(rustc_private)]
 
 #[macro_use]
 extern crate syn;
@@ -10,10 +8,6 @@ use syn::{BinOp, Expr, ExprBinary, ExprGroup, ExprLit, Lit};
 extern crate proc_macro2;
 use proc_macro2::*;
 
-#[macro_use]
-mod macros;
-
-mod common;
 mod features;
 
 fn expr<T: Into<Expr>>(t: T) -> Expr {

--- a/tests/test_grouping.rs
+++ b/tests/test_grouping.rs
@@ -1,3 +1,4 @@
+#![cfg(not(syn_disable_nightly_tests))]
 #![recursion_limit = "1024"]
 #![feature(rustc_private)]
 

--- a/tests/test_precedence.rs
+++ b/tests/test_precedence.rs
@@ -1,3 +1,4 @@
+#![cfg(not(syn_disable_nightly_tests))]
 #![recursion_limit = "1024"]
 #![feature(rustc_private)]
 

--- a/tests/test_round_trip.rs
+++ b/tests/test_round_trip.rs
@@ -1,3 +1,4 @@
+#![cfg(not(syn_disable_nightly_tests))]
 #![recursion_limit = "1024"]
 #![feature(rustc_private)]
 

--- a/tests/zzz_stable.rs
+++ b/tests/zzz_stable.rs
@@ -1,0 +1,23 @@
+#![cfg(syn_disable_nightly_tests)]
+
+extern crate colored;
+
+use colored::Colorize;
+
+const MSG: &str = "
+‖
+‖   WARNING:
+‖   This is not a nightly compiler so not all tests were able to
+‖   run. Syn includes tests that compare Syn's parser against the
+‖   compiler's parser, which requires access to unstable libsyntax
+‖   data structures and a nightly compiler.
+‖
+";
+
+#[test]
+fn notice() {
+    panic!(MSG
+        .replace("WARNING", &"WARNING".bold().to_string())
+        .yellow()
+        .to_string());
+}

